File: medium/Two-Pointers/squares-of-a-sorted-array/sorted_squares.py
from typing import List

class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        """
        Generate squares of sorted array in non-decreasing order.
        Approach: Left & Right Two Pointers
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        n = len(nums)
        res = [0] * n
        left = 0
        right = n - 1
        idx = n - 1
        
        while left <= right:
            left_sq = nums[left] ** 2
            right_sq = nums[right] ** 2
            if left_sq > right_sq:
                res[idx] = left_sq
                left += 1
            else:
                res[idx] = right_sq
                right -= 1
            idx -= 1
        return res

File: medium/Two-Pointers/squares-of-a-sorted-array/sorted_squares.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        /*
         * Compute square of each element and return sorted array.
         * Approach: Left Right Two Pointers from both ends
         * Time Complexity: O(n)
         * Space Complexity: O(n)
         */
        int n = nums.size();
        vector<int> res(n);
        int left = 0, right = n - 1, idx = n - 1;
        
        while(left <= right){
            int leftSq = nums[left] * nums[left];
            int rightSq = nums[right] * nums[right];
            if(leftSq > rightSq){
                res[idx--] = leftSq;
                left++;
            }else{
                res[idx--] = rightSq;
                right--;
            }
        }
        return res;
    }
};
